src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,625 @@
+  std::string ext_new_;
+  std::string module_dir_;
+
+ public:
+  ModuleConfig() :
+#if defined(__APPLE__) && defined(__MACH__)
+    is_darwin_(true),
+    java_md_("darwin"),
+#else
+    is_darwin_(false),
+    java_md_("linux"),
+#endif
+    ext_cpp_(".cpp"),
+    ext_lib_(is_darwin_ ? ".dylib" : ".so"),
+    ext_mak_(".mak"),
+    ext_new_(".new"),
+    module_dir_(get_module_dir()) {
+  }
+  
+  std::string get_lib_name(const std::string& key) {
+    std:: stringstream ss;
+    ss << module_dir_ << "/hm_" << key << ext_lib_;
+    return ss.str();
+  }
+  
+  std::string get_new_name(const std::string& key) {
+    std:: stringstream ss;
+    ss << module_dir_ << "/hm_" << key << ext_new_;
+    return ss.str();
+  }
+  
+  void ensure_module_dir_exists() {
+    int rc = ::access(module_dir_.c_str(), R_OK);
+    if (rc < 0) { // create it
+      rc = ::mkdir(module_dir_.c_str(), 0666);
+      if (rc < 0) perror(module_dir_.c_str());
+      rc = ::chmod(module_dir_.c_str(), 0755);
+    }
+  }
+};
+
+ModuleConfig config;
+
+// module_table is used to ensure that several Spark worker threads will get
+// shared_ptr's to a single NativeModule, rather than each having their
+// own NativeModule.
+
+std::unordered_map<std::string, std::weak_ptr<NativeModule>> module_table;
+
+} // end anon
+
+// ModuleBuilder deals with compiling/linking source code to a DLL,
+// and providing the binary DLL as an Array[Byte] which can be broadcast
+// to all workers.
+
+class ModuleBuilder {
+ private:
+  std::string options_;
+  std::string source_;
+  std::string include_;
+  std::string key_;
+  std::string hm_base_;
+  std::string hm_mak_;
+  std::string hm_cpp_;
+  std::string hm_new_;
+  std::string hm_lib_;
+  
+public:
+  ModuleBuilder(
+    const std::string& options,
+    const std::string& source,
+    const std::string& include,
+    const std::string& key
+  ) :
+    options_(options),
+    source_(source),
+    include_(include),
+    key_(key) {
+    // To start with, put dynamic code in $HOME/hail_modules
+    auto base = (config.module_dir_ + "/hm_") + key_;
+    hm_base_ = base;
+    hm_mak_ = (base + config.ext_mak_);
+    hm_cpp_ = (base + config.ext_cpp_);
+    hm_new_ = (base + config.ext_new_);
+    hm_lib_ = (base + config.ext_lib_);
+  }
+  
+  virtual ~ModuleBuilder() { }
+  
+private:
+  void write_cpp() {
+    FILE* f = fopen(hm_cpp_.c_str(), "w");
+    if (!f) { perror("fopen"); return; }
+    fwrite(source_.data(), 1, source_.length(), f);
+    fclose(f);
+  }
+  
+  void write_mak() {
+    // When we replace foo.new, or foo.{so,dylib}, that must be done
+    // with an atomic-rename operation which guarantees not to leave any
+    // window when there is no foo.new.  On Linux, "mv -f" is atomic, but
+    // on MacOS it isn't.  After some experimentation, we now use perl's
+    // rename command, with the belief that on POSIX-compatible systems
+    // this will be implemented as a rename() system-call, specified thus:
```

Comment:
I think this `.new` stuff is no longer necessary.  In particular, I'd be very happy to get rid of dependence on perl.

Reply:
I think what complicates things a little right now the "force_build" option, which means that a
master module will rebuild the lib even if it already exists.  When that interacts with a test environment where we're running several tests in different threads, it's possible for a module.so which is still being used by some test to be rebuilt and overwritten by another test - and in that situation I would be nervous about eliminating the atomic replace.

But that could also be an argument for just removing the "force_build" option - especially since the module key now depends on checksum of header files and output of "c++ --version".

Reply:
Yeah, I'd just remove force_build.